### PR TITLE
Encoding fixes and cleanup

### PR DESCRIPTION
**Pull request by eregon**

### CODE REVIEW

File: src/launcher/java/org/truffleruby/launcher/options/CommandLineParser.java [hunk 1]
```diff
@@ -559,7 +559,7 @@ private void processEncodingOption(String value) throws CommandLineException {
         }
 
         if (encodings.size() >= 1) {
-            config.setOption(OptionsCatalog.EXTERNAL_ENCODING, encodings.get(1));
+            config.setOption(OptionsCatalog.EXTERNAL_ENCODING, encodings.get(0));
```
Comment (nirvdrum) [inline src/launcher/java/org/truffleruby/launcher/options/CommandLineParser.java]: Ugh.

**Comment by eregon** [comment]
@nirvdrum I pushed a few more cleanup commits, could you take a look?

File: src/launcher/java/org/truffleruby/launcher/options/CommandLineParser.java (hunk 1 above)
Comment (eregon) [reply src/launcher/java/org/truffleruby/launcher/options/CommandLineParser.java]: I was curious: https://github.com/oracle/truffleruby/commit/f952a08edb57524d67884f2dead34abe58a74bb8#r20633204 introduced it. And I even noticed, @chrisseaton ;)

File: src/main/ruby/core/encoding.rb [hunk 2]
```diff
@@ -41,7 +41,7 @@ class << self
         map[key] = [nil, index]
       end
 
-      Truffle::EncodingOperations.each_alias do |alias_name, index|
+      Truffle.invoke_primitive :encoding_each_alias, -> alias_name, index do
```
[the remark this answers is not in the document]
Comment (eregon) [reply src/main/ruby/core/encoding.rb]: Yes, primitives don't support `needsBlock` currently and reading the block manually seems more hacky than just passing a lambda explicitly.

File: src/main/java/org/truffleruby/core/encoding/EncodingNodes.java [hunk 3]
```diff
@@ -440,6 +440,36 @@ public boolean isUnicode(DynamicObject encoding) {
 
     }
 
+    @Primitive(name = "encoding_get_default_encoding", needsSelf = false)
+    public abstract static class GetDefaultEncodingNode extends PrimitiveArrayArgumentsNode {
+
+        @Specialization(guards = "isRubyString(name)")
```
[the remark this answers is not in the document]
Comment (eregon) [reply src/main/java/org/truffleruby/core/encoding/EncodingNodes.java]: Done.

File: src/main/java/org/truffleruby/core/encoding/EncodingNodes.java [hunk 4]
```diff
@@ -494,6 +494,39 @@ private Encoding getEncoding(String name) {
...
+    public abstract static class SetDefaultExternalNode extends PrimitiveArrayArgumentsNode {
+
+        @Specialization(guards = "isRubyEncoding(encoding)")
+        public DynamicObject defaultExternalEncoding(DynamicObject encoding) {
+            getContext().getEncodingManager().setDefaultExternalEncoding(EncodingOperations.getEncoding(encoding));
+            return encoding;
+        }
+
+        @Specialization(guards = "isNil(nil)")
+        public DynamicObject defaultExternal(Object nil) {
```
[the remark this answers is not in the document]
Comment (eregon) [reply src/main/java/org/truffleruby/core/encoding/EncodingNodes.java]: Done.

File: src/main/java/org/truffleruby/core/encoding/EncodingNodes.java [hunk 5]
```diff
@@ -494,6 +494,39 @@ private Encoding getEncoding(String name) {
...
+    public abstract static class SetDefaultInternalNode extends PrimitiveArrayArgumentsNode {
+
+        @Specialization(guards = "isRubyEncoding(encoding)")
+        public DynamicObject defaultInternal(DynamicObject encoding) {
+            getContext().getEncodingManager().setDefaultInternalEncoding(EncodingOperations.getEncoding(encoding));
+            return encoding;
+        }
+
+        @Specialization(guards = "isNil(encoding)")
+        public DynamicObject defaultInternal(Object encoding) {
```
[the remark this answers is not in the document]
Comment (eregon) [reply src/main/java/org/truffleruby/core/encoding/EncodingNodes.java]: Done.

**Pull request merged**